traits/observers/_dict_item_observer.py

```diff
@@ -0,0 +1,205 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from traits.observers._dict_change_event import dict_event_factory
+from traits.observers._i_observer import IObserver
+from traits.observers._observe import add_or_remove_notifiers
+from traits.observers._observer_change_notifier import ObserverChangeNotifier
+from traits.observers._trait_event_notifier import TraitEventNotifier
+from traits.trait_dict_object import TraitDict
+
+
+@IObserver.register
+class DictItemObserver:
+    """ Observer for observing mutations on a dict.
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    optional : boolean
+        If False, this observer will complain if the incoming object is not
+        an observable dict. If True and the incoming object is not a dict, this
+        observer will do nothing. Useful for the 'items' keyword in the text
+        parser, where the source container type is ambiguous.
+    """
+
+    def __init__(self, *, notify, optional):
+        self.notify = notify
+        self.optional = optional
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self).__name__, self.notify, self.optional))
+
+    def __eq__(self, other):
+        """ Return true if this observer is equal to the given one."""
+        return (
+            type(self) is type(other)
```

Comment:
We're still inconsistent across the codebase about using `type` (e.g., as here) versus using `.__class__` (e.g., in the `__repr__` for `DictChangeEvent`. They're subtly different, and some of the new typing machinery being introduced in recent Python versions is enlarging that difference (e.g., in Python 3.9, `list[int]` has a different `__class__` from its `type`).

I _think_ we actually want to use `__class__` consistently throughout the codebase, despite the general rule that one shouldn't generally use dunder attributes and methods directly as a consumer.

Not something that should be changed in this PR; just something to think about.